File: main_client.cc
#include "guide.grpc.pb.h"
#include <iostream>
#include <string>
#include <memory>
#include <grpc/grpc.h>
#include <grpcpp/channel.h>
#include <grpcpp/client_context.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/security/credentials.h>

int main() {
  std::string server_addr = "localhost:50051";
  std::shared_ptr<grpc::Channel> channel =
      grpc::CreateChannel(server_addr, grpc::InsecureChannelCredentials());
  std::unique_ptr<Guide::Stub> stub = Guide::NewStub(channel);

  while (true) {
    grpc::ClientContext context;
    grpc::Status status;
    CarState car_state;
    Step step;
    int car_id = 0;
    int cur_row_idx = 0;
    int cur_col_idx = 0;
    int goal_row_idx = 0;
    int goal_col_idx = 0;

    std::cin >> car_id >> cur_row_idx >> cur_col_idx
             >> goal_row_idx >> goal_col_idx;
    car_state.set_car_id(car_id);
    car_state.set_cur_row_idx(cur_row_idx);
    car_state.set_cur_col_idx(cur_col_idx);
    car_state.set_goal_row_idx(goal_row_idx);
    car_state.set_goal_col_idx(goal_col_idx);

    status = stub->GetNextStep(&context, car_state, &step);

    if (status.ok())
      std::cout << step.step_code() << std::endl;
    else
      std::cout << status.error_message() << std::endl;
  }

  return 0;
}

File: main_client.py
import grpc
import guide_pb2
import guide_pb2_grpc


def main():
    channel = grpc.insecure_channel('localhost:50051')
    stub = guide_pb2_grpc.GuideStub(channel)
    
    while True:
        line = input()
        tokens = line.split()
        car_id = int(tokens[0])
        cur_row_idx = int(tokens[1])
        cur_col_idx = int(tokens[2])
        goal_row_idx = int(tokens[3])
        goal_col_idx = int(tokens[4])
        car_state = guide_pb2.CarState(
            car_id=car_id,
            cur_row_idx=cur_row_idx,
            cur_col_idx=cur_col_idx,
            goal_row_idx=goal_row_idx,
            goal_col_idx=goal_col_idx,
        )

        try:
            step = stub.GetNextStep(car_state)
            print(step)
        except grpc.RpcError as e:
            print(e.code())


if __name__ == '__main__':
    main()
